### PR TITLE
Allow custom AWS tags in iamrole annotation

### PR DESCRIPTION
**Pull request by nrosenberg1**
**Could you share the solution in high level?**
1. Allow specifying custom tags to be applied to Iamrole in AWS along with default tags
2. User can annotate Iamrole with key `"iammanager.keikoproj.io/tags"` with value in format `"key1=value1;;key2=value2"` and controller will tag AWS role accordingly

**Could you share the test results?**
Screen Shot 2022-03-14 at 3 20 56 PM
Screen Shot 2022-03-09 at 12 11 05 PM

### CODE REVIEW

File: controllers/iamrole_controller.go [hunk 1]
```diff
@@ -298,6 +298,21 @@ func (r *IamroleReconciler) ConstructCreateIAMRoleInput(ctx context.Context, iam
 		tags["Cluster"] = config.Props.ClusterName()
 	}
 
+	// Custom tags value should be a string of comma seperated key/value pairs
+	// example annotation: "iamroles.iammanager.keikoproj.io/tags": "key1=value1,key2=value2"
+	if customTagsString, ok := iamRole.GetAnnotations()["iamroles.iammanager.keikoproj.io/tags"]; ok {
```
Comment (mnkg561) [inline controllers/iamrole_controller.go:303]: There is a common utility function  to read the annotation. Please use that

File: pkg/validation/validate.go [hunk 2]
```diff
@@ -168,6 +174,34 @@ func CompareAssumeRolePolicy(ctx context.Context, request string, target string)
 	return true
 }
 
+//CompareTags compares tags from request and response
+func CompareTags(ctx context.Context, request map[string]string, target []*iam.Tag) bool {
```
Comment (mnkg561) [inline pkg/validation/validate.go:178]: Does this handle the default tags in the comparison? Default tags(managedBy, Namespace, cluster) is not coming in the request annotation so it is always returns that there is a diff

**Review by mnkg561 — COMMENTED** [review]
Also, Make sure you are not allowing people to overwrite default tags like namespace, cluster, managedBy using annotations. Those are reserved tags and should not be allowed to edited by the user

File: controllers/iamrole_controller.go (hunk 1 above)
Comment (nrosenberg1) [reply controllers/iamrole_controller.go:303]: Sure, will change this

File: pkg/validation/validate.go (hunk 2 above)
Comment (nrosenberg1) [reply pkg/validation/validate.go:178]: This is checking roleInput tags so default tags are included in request map, unit tests validate the same

**Comment by nrosenberg1** [comment]
> Also, Make sure you are not allowing people to overwrite default tags like namespace, cluster, managedBy using annotations. Those are reserved tags and should not be allowed to edited by the user

Sure, I will add this validation

**Review by mnkg561 — APPROVED** [review]
If you already taken care of protecting reserved tags, its good to go

**Comment by nrosenberg1** [comment]
@mnkg561 is travis-ci status check no longer needed since moving to github actions? https://github.com/keikoproj/iam-manager/pull/85

**Comment by mnkg561** [comment]
There is no travis.yaml file in the repo. I also removed travis web hook just now. Try with a dummy commit and see whether it gets refreshed

**Comment by nrosenberg1** [comment]
@mnkg561 looks like dummy commit didn't refresh it, guess I will just administrator merge

**Pull request merged**